cpp/ql/src/Security/CWE/CWE-843/TypeConfusion.ql

```diff
@@ -0,0 +1,228 @@
+/**
+ * @name Type confusion
+ * @description Casting a value to an incompatible type can lead to undefined behavior.
+ * @kind path-problem
+ * @problem.severity warning
+ * @security-severity 9.3
+ * @precision medium
+ * @id cpp/type-confusion
+ * @tags security
+ *       external/cwe/cwe-843
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.new.DataFlow
+import BadFlow::PathGraph
+
+/**
+ * Holds if `f` is a field located at byte offset `offset` in `c`.
+ *
+ * Note that predicate is recursive, so that given the following:
+ * ```cpp
+ * struct S1 {
+ *   int a;
+ *   void* b;
+ * };
+ *
+ * struct S2 {
+ *   S1 s1;
+ *   char c;
+ * };
+ * ```
+ * both `hasAFieldWithOffset(S2, s1, 0)` and `hasAFieldWithOffset(S2, a, 0)`
+ * holds.
+ */
+predicate hasAFieldWithOffset(Class c, Field f, int offset) {
+  // Let's not deal with bit fields for now.
+  not f instanceof BitField and
+  (
+    // Base case: `f` is a field in `c`.
+    f = c.getAField() and
+    offset = f.getByteOffset()
+    or
+    // Otherwise, we find the struct that is a field of `c` which then has
+    // the field `f` as a member.
+    exists(Field g |
+      g = c.getAField() and
+      // Find the field with the largest offset that's less than offset. That's
+      // the struct we need to search recursively.
+      g =
+        max(Field cand, int candOffset |
+          cand = c.getAField() and
+          candOffset = cand.getByteOffset() and
+          offset >= candOffset
+        |
+          cand order by candOffset
+        ) and
+      hasAFieldWithOffset(g.getUnspecifiedType(), f, offset - g.getByteOffset())
+    )
+  )
+}
+
+/** Holds if `f` is the last field of its declaring class. */
+predicate lastField(Field f) {
+  exists(Class c | c = f.getDeclaringType() |
+    f =
+      max(Field cand, int byteOffset |
+        cand.getDeclaringType() = c and byteOffset = f.getByteOffset()
+      |
+        cand order by byteOffset
+      )
+  )
+}
+
+/**
+ * Holds if there exists a field in `c2` at offset `offset` that's compatible
+ * with `f1`.
+ */
+bindingset[f1, offset, c2]
+pragma[inline_late]
+predicate hasCompatibleFieldAtOffset(Field f1, int offset, Class c2) {
+  exists(Field f2 | hasAFieldWithOffset(c2, f2, offset) |
+    f1.getUnspecifiedType().getSize() = f2.getUnspecifiedType().getSize()
+    or
+    lastField(f1) and
+    f1.getUnspecifiedType().getSize() <= f2.getUnspecifiedType().getSize()
```

Comment:
That's one case, yeah. Another one is something like:
```cpp
struct UInt64 { unsigned long u64; };
struct UInt8 { unsigned char u8; };

void use(...);

void test() {
  void* u64 = new UInt64;
  // ...
  UInt8* u8 = (UInt8*)u64;
  use(u8->u8);
}
```
where you pretend that the final value in the struct is a smaller type than it actually is. This is fine since there's no "next" value that's then not "aligned". For example, if you'd have something like:
```cpp
struct UInt64 { unsigned long u64; };
struct UInt8 { unsigned char u8; void* p; };

void use(...);

void test() {
  void* u64 = new UInt64;
  // ...
  UInt8* u8 = (UInt8*)u64;
  use(u8->u8);
  use(u8->p); // This is terribly dangerous!!!
}
```
then the value of `p` will be garbage. I've added both of these examples to the tests.